### PR TITLE
Add macro to generate main with 2 flavors: regular and VLD leak checks

### PR DESCRIPTION
**Pull request by dcristoloveanu**
Add macro to generate main with 2 flavors: regular and VLD leak checks

### CODE REVIEW

File: inc/ctrs_leak_check.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#ifndef CTRS_LEAK_CHECK_H
+#define CTRS_LEAK_CHECK_H
+
+#ifdef __cplusplus
+#include <cstddef>
+#else
+#include <stddef.h>
+#endif
+
+#include "testrunnerswitcher.h"
```
Comment (anporumb) [inline inc/ctrs_leak_check.h]: >testrunnerswitcher [](start = 10, length = 18)

This file (ctrs_leak_check) should include vld.h

Comment (anporumb) [reply inc/ctrs_leak_check.h]: LIkely under VLD_OPT_REPORT_TO_STDOUT

---
In reply to: [291298918](https://github.com/Azure/azure-c-testrunnerswitcher/pull/19#discussion_r291298918) [](ancestors = 291298918)

File: inc/ctrs_leak_check.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+#ifdef VLD_OPT_REPORT_TO_STDOUT
+// MAIN_RUN_TEST_SUITE_WITH_LEAK_CHECK macro with VLD leak checks
+#define MAIN_RUN_TEST_SUITE_WITH_LEAK_CHECK(test_suite) \
+int main(void) \
+{ \
+    size_t failedTestCount = 0; \
+    VLD_UINT initial_leak_count = VLDGetLeaksCount(); \
+    RUN_TEST_SUITE(test_suite, failedTestCount); \
```
Comment (anporumb) [inline inc/ctrs_leak_check.h]: >RUN_TEST_SUITE [](start = 4, length = 14)

But why not have the vld.h and its glorious functionality under RUN_TEST_SUITE macro?

File: inc/ctrs_leak_check.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    failedTestCount = (failedTestCount > 0) ? failedTestCount : -(int)(VLDGetLeaksCount() - initial_leak_count); \
+    return failedTestCount; \
+}
+#else
+// MAIN_RUN_TEST_SUITE_WITH_LEAK_CHECK macro without VLD leak checks
+#define MAIN_RUN_TEST_SUITE_WITH_LEAK_CHECK(test_suite) \
```
Comment (anporumb) [inline inc/ctrs_leak_check.h]: >MAIN_RUN_TEST_SUITE_WITH_LEAK_CHECK [](start = 8, length = 35)

This macro doesn't really allow anymore to run 2 test suites in the same file, does it?

File: inc/ctrs_leak_check.h (hunk 1 above)
Comment (dcristoloveanu) [reply inc/ctrs_leak_check.h]: I'll add it. Truth be told this is piggybacking the fact that vld.h was included. But maybe in the future we should have an explicit switch for this. But maybe in the future :-). Let's not do all the improvements in one day.

---
In reply to: [291299367](https://github.com/Azure/azure-c-testrunnerswitcher/pull/19#discussion_r291299367) [](ancestors = 291299367,291298918)

File: inc/ctrs_leak_check.h (hunk 2 above)
Comment (dcristoloveanu) [reply inc/ctrs_leak_check.h]: Initially this was a macro in the consumer repo (in test_helpers). But then test_helpers would need to be added to all CMakeLists. So I decided to move the macro to ctestrunnerswitcher.

But I agree, having it under RUN_TEST_SUITE is much better.

---
In reply to: [291299922](https://github.com/Azure/azure-c-testrunnerswitcher/pull/19#discussion_r291299922) [](ancestors = 291299922)

**Comment by dcristoloveanu** [comment]
The leak check will be moved to ctest actually, thus closing this PR.

**Pull request closed without merging**